cmd/snap-confine/ns-support.c

```diff
@@ -306,10 +308,172 @@ static bool should_discard_current_ns(dev_t base_snap_dev)
+	sc_must_snprintf(fname, sizeof fname, "%s/%s/current",
+			 SNAP_MOUNT_DIR, base_snap_name);
+	if (readlink(fname, base_snap_rev, sizeof base_snap_rev) < 0) {
+		die("cannot read revision of base snap %s", fname);
+	}
+	if (base_snap_rev[sizeof base_snap_rev - 1] != '\0') {
+		die("cannot use symbolic link %s - value is too long", fname);
+	}
+	// Find the device that is backing the current revision of the base snap.
+	base_snap_dev = find_base_snap_device(base_snap_name, base_snap_rev);
+
+	// Check if we are running on classic. Do it here because we will always
+	// (seemingly) run on a core system once we are inside a mount namespace.
+	bool is_classic = is_running_on_classic_distribution();
+
+	// Store the PID of this process. This is done instead of calls to
+	// getppid() below because then we can reliably track the PID of the
+	// parent even if the child process is re-parented.
+	parent = getpid();
+
+	// Create an eventfd for the communication with the child.
+	event_fd = eventfd(0, EFD_CLOEXEC);
+	if (event_fd < 0) {
+		die("cannot create eventfd for communication with inspection process");
+	}
+	// Fork a child, it will do the inspection for us.
+	child = fork();
+	if (child < 0) {
+		die("cannot fork support process for namespace inspection");
+	}
+
+	if (child == 0) {
+		// This is the child process which will inspect the mount namespace.
+		//
+		// Configure the child to die as soon as the parent dies. In an odd
+		// case where the parent is killed then we don't want to complete our
+		// task or wait for anything.
+		if (prctl(PR_SET_PDEATHSIG, SIGINT, 0, 0, 0) < 0) {
+			die("cannot set parent process death notification signal to SIGINT");
+		}
+		// Check that parent process is still alive. If this is the case then
+		// we can *almost* reliably rely on the PR_SET_PDEATHSIG signal to wake
+		// us up from eventfd_read() below. In the rare case that the PID
+		// numbers overflow and the now-dead parent PID is recycled we will
+		// still hang forever on the read from eventfd below.
+		debug("ensuring that parent process is still alive");
+		if (kill(parent, 0) < 0) {
+			switch (errno) {
+			case ESRCH:
+				debug("parent process has already terminated");
+				abort();
+			default:
+				die("cannot ensure that parent process is still alive");
+				break;
+			}
+		}
+
+		debug("joining the namespace that we are about to probe");
+		// Move to the mount namespace of the snap we're trying to inspect.
+		if (setns(mnt_fd, CLONE_NEWNS) < 0) {
+			die("cannot join the mount namespace in order to inspect it");
+		}
+		// Check if the namespace needs to be discarded.
+		//
+		// TODO: enable this for core distributions. This is complex because on
+		// core the rootfs is mounted in initrd and is _not_ changed (no
+		// pivot_root) and the base snap is again mounted (2nd time) by
+		// systemd. This makes us end up in a situation where the outer base
+		// snap will never match the rootfs inside the mount namespace.
```

Comment:
It seems weird to commit this for only classic. What is the plan here? Why is it ok to ignore core for now? Should we express that in the comment?

Reply:
Uh, this is a rat's nest I'm afraid. I did it for classic only because that is easy to land and because core reboots anyway.

The real reason is that core is never correct in this case because / is never the core snap as it is mounted in /snap/core/current/ because the real / is mounted from initramfs. In reality on all-snap systems core and kernel snaps are always mounted twice.

The logic to handle this correctly is different and I wanted to do it in a followup.